File: src/main/Python/nowcoder/S33_UglyNumber.py
class UglyNumber:
    def UglyNumber(self, index):
        arr = [1]
        if index < 7:
            return index
        p2, p3, p5 = 0, 0, 0
        while len(arr) < index:
            v2 = arr[p2] * 2
            v3 = arr[p3] * 3
            v5 = arr[p5] * 5
            mini = min(v2, v3, v5)
            arr.append(mini)
            if v2 == mini: p2 += 1
            if v3 == mini: p3 += 1
            if v5 == mini: p5 += 1
        return arr[-1]
if __name__ == '__main__':
    test = UglyNumber()
    print(test.UglyNumber(100))

File: src/main/java/nowcoder/S33_UglyNumber.java
package nowcoder;

import java.util.ArrayList;

/**
 * @author LiuZhiguo
 * @date 2020/1/28 23:39
 * 把只包含质因子2、3和5的数称作丑数（Ugly Number）。例如6、8都是丑数，
 * 但14不是，因为它包含质因子7。 习惯上我们把1当做是第一个丑数。求按从小到大的顺序的第N个丑数。
 */
public class S33_UglyNumber {
    public int uglyNumber(int index){
        if (index < 7) return index;
        ArrayList<Integer> arr = new ArrayList<Integer>();
        arr.add(1); //将1加入丑数数组中
        int p2 = 0, p3 = 0, p5 = 0;
        while (arr.size() < index) {
            int v2 = arr.get(p2) * 2;
            int v3 = arr.get(p3) * 3;
            int v5 = arr.get(p5) * 5;
            int min = Math.min(v2, Math.min(v3, v5));
            arr.add(min);
            if (v2 == min) p2++;
            if (v3 == min) p3++;
            if (v5 == min) p5++;
        }
        return arr.get(arr.size()-1);
    }
    public static void main(String[] args){
        S33_UglyNumber s33 = new S33_UglyNumber();
        System.out.println(s33.uglyNumber(100));
    }
}
